CDI interceptors fail in Weblogic

The CDI interceptors do not follow the CDI spec:

> (section 2.5 of Interceptor Specs)
> 
> Lifecycle callback interceptor methods defined on an interceptor class must have one of the following signatures:
> 
> void (InvocationContext)
> Object (InvocationContext) throws Exception

The CDI adaptors (for example org.jsr107.ri.annotations.cdi.CacheRemoveAllInterceptor) have this signature instead: 

> @AroundInvoke
> public Object cacheRemoveAll(InvocationContext invocationContext) throws Throwable

Please change 'Throwable' to 'Exception' so that this weblogic error does not come anymore:

> `[ERROR] weblogic.ejb.container.compliance.ComplianceException: AroundInvoke method cacheRemoveAll in class org.jsr107.ri.annotations.cdi.CacheRemoveAllInterceptor must have the signature: 'Object <METHOD(InvocationContext)> throws Exception'`
> `[ERROR] at weblogic.ejb.container.compliance.BusinessMethodInterceptorChecker.validateAroundInvokeOrTimeoutMethod(BusinessMethodInterceptorChecker.java:154)`
> `[ERROR] at weblogic.ejb.container.compliance.BusinessMethodInterceptorChecker.checkAroundInvokeOrTimeoutMethods(BusinessMethodInterceptorChecker.java:70)`
> `[ERROR] at weblogic.ejb.container.compliance.InterceptorChecker.checkAroundInvokeMethods(InterceptorChecker.java:61)`
> `[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)`
> `[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)`
> `[ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)`
> `[ERROR] at java.lang.reflect.Method.invoke(Method.java:497)`
> `[ERROR] at weblogic.ejb.container.compliance.EJBComplianceChecker.check(EJBComplianceChecker.java:100)`

Affects both Weblogic 12.1.3 and 12.2.1.

--- a/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CachePutInterceptor.java
+++ b/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CachePutInterceptor.java
@@ -44,11 +44,19 @@ public class CachePutInterceptor extends AbstractCachePutInterceptor<InvocationC
   /**
    * @param invocationContext The intercepted invocation
    * @return The result from {@link InvocationContext#proceed()}
-   * @throws Throwable likely {@link InvocationContext#proceed()} threw an exception
+   * @throws Exception likely {@link InvocationContext#proceed()} threw an exception
    */
   @AroundInvoke
-  public Object cachePut(InvocationContext invocationContext) throws Throwable {
-    return this.cachePut(this.lookup, invocationContext);
+  public Object cachePut(InvocationContext invocationContext) throws Exception {
+    try {
+        return this.cachePut(this.lookup, invocationContext);
+    } catch (Throwable e) {
+        if (e instanceof Exception) {
+            throw (Exception) e;
+        } else {
+            throw new Exception(e);
+        }
+    }
   }
 
   /* (non-Javadoc)

--- a/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheRemoveAllInterceptor.java
+++ b/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheRemoveAllInterceptor.java
@@ -42,11 +42,19 @@ public class CacheRemoveAllInterceptor extends AbstractCacheRemoveAllInterceptor
   /**
    * @param invocationContext The intercepted invocation
    * @return The result from {@link InvocationContext#proceed()}
-   * @throws Throwable likely {@link InvocationContext#proceed()} threw an exception
+   * @throws Exception likely {@link InvocationContext#proceed()} threw an exception
    */
   @AroundInvoke
-  public Object cacheRemoveAll(InvocationContext invocationContext) throws Throwable {
-    return this.cacheRemoveAll(this.lookup, invocationContext);
+  public Object cacheRemoveAll(InvocationContext invocationContext) throws Exception {
+    try {
+        return this.cacheRemoveAll(this.lookup, invocationContext);
+    } catch (Throwable e) {
+        if (e instanceof Exception) {
+            throw (Exception) e;
+        } else {
+            throw new Exception(e);
+        }
+    }
   }
 
   /* (non-Javadoc)

--- a/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheRemoveEntryInterceptor.java
+++ b/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheRemoveEntryInterceptor.java
@@ -44,11 +44,19 @@ public class CacheRemoveEntryInterceptor extends AbstractCacheRemoveEntryInterce
   /**
    * @param invocationContext The intercepted invocation
    * @return The result from {@link InvocationContext#proceed()}
-   * @throws Throwable likely {@link InvocationContext#proceed()} threw an exception
+   * @throws Exception likely {@link InvocationContext#proceed()} threw an exception
    */
   @AroundInvoke
-  public Object cacheRemoveEntry(InvocationContext invocationContext) throws Throwable {
-    return this.cacheRemoveEntry(this.lookup, invocationContext);
+  public Object cacheRemoveEntry(InvocationContext invocationContext) throws Exception {
+    try {
+        return this.cacheRemoveEntry(this.lookup, invocationContext);
+    } catch (Throwable e) {
+        if (e instanceof Exception) {
+            throw (Exception) e;
+        } else {
+            throw new Exception(e);
+        }
+    }
   }
 
   /* (non-Javadoc)

--- a/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheResultInterceptor.java
+++ b/cache-annotations-ri/cache-annotations-ri-cdi/src/main/java/org/jsr107/ri/annotations/cdi/CacheResultInterceptor.java
@@ -43,11 +43,19 @@ public class CacheResultInterceptor extends AbstractCacheResultInterceptor<Invoc
   /**
    * @param invocationContext The intercepted invocation
    * @return The result from {@link InvocationContext#proceed()}
-   * @throws Throwable likely {@link InvocationContext#proceed()} threw an exception
+   * @throws Exception likely {@link InvocationContext#proceed()} threw an exception
    */
   @AroundInvoke
-  public Object cacheResult(InvocationContext invocationContext) throws Throwable {
-    return this.cacheResult(this.lookup, invocationContext);
+  public Object cacheResult(InvocationContext invocationContext) throws Exception {
+    try {
+        return this.cacheResult(this.lookup, invocationContext);
+    } catch (Throwable e) {
+        if (e instanceof Exception) {
+            throw (Exception) e;
+        } else {
+            throw new Exception(e);
+        }
+    }
   }
 
   /* (non-Javadoc)

